src/lxml/apihelpers.pxi

```diff
@@ -1582,6 +1582,25 @@ cdef bint _isFilePath(const_xmlChar* c_path):
     # assume it's a relative path
     return REL_FILE_PATH
 
+cdef object _getFSPathOrObject(object obj):
+    u"""
+    Get the __fspath__ attribute of an object
+    if it exists otherwise the original object
+    is returned
+    """
+    if python.PY_VERSION_HEX >= 0x03060000:
+        try:
+            fspath = python.PY_FSPath(obj)
+        except TypeError:
+            return obj
+        return fspath
+    elif hasattr(obj, '__fspath__'):
+        fspath = obj.__fspath__
+        if callable(fspath):
+            return fspath()
+        raise ValueError('__fspath__ is not callable')
+    return obj
```

Comment:
I think we can keep the `_isString()` check, because it's so incredibly cheap.
For Py<3.6, since users probably weren't expecting `__fspath__` to become special, I think ignoring anything we find that's not callable is better than shouting at them.
Finally, I consider `hasattr()` followed by an attribute access an anti-pattern. Why look up the attribute twice? If you want to read the attribute, do so. We can use the 3-args `getattr()` in this case, to avoid the likely `AttributeError` if possible.
```suggestion
cdef object _NO_FSPATH = object()

cdef object _getFSPathOrObject(object obj):
    """
    Get the __fspath__ attribute of an object if it exists.
    Otherwise, the original object is returned.
    """
    if _isString(obj):
        return obj
    if python.PY_VERSION_HEX >= 0x03060000:
        try:
            return python.PY_FSPath(obj)
        except TypeError:
            return obj
    fspath = getattr(obj, '__fspath__', _NO_FSPATH)
    if fspath is not _NO_FSPATH and callable(fspath):
        return fspath()
    return obj
```